File: app/src/main/java/com/my/televip/obfuscate/struct/ClientObfuscationData.java
package com.my.televip.obfuscate.struct;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class ClientObfuscationData {

    private final Map<String, String> classMap = new HashMap<>();
    private final Map<String, String> fieldMap = new HashMap<>();
    private final Map<String, String> methodMap = new HashMap<>();
    private final Map<String, String> methodAliasMap = new HashMap<>();

    public static ClientObfuscationData parse(String clientJson, String clientAliasJson) {
        ClientObfuscationData data = new ClientObfuscationData();
        try {
            JSONObject root = new JSONObject(clientJson);
            data.parseClasses(root.optJSONArray("classes"));
            data.parseFields(root.optJSONArray("fields"));
            data.parseMethods(root.optJSONArray("methods"));
            JSONObject aliasRoot = new JSONObject(clientAliasJson);
            data.parseMethodAlias(aliasRoot.optJSONObject("methodAlias"));
        } catch (Throwable t) {
            t.printStackTrace();
        }
        return data;
    }

    private void parseClasses(JSONArray arr) {
        if (arr == null) return;
        for (int i = 0; i < arr.length(); i++) {
            JSONObject o = arr.optJSONObject(i);
            if (o == null) continue;
            String original = o.optString("o", null);
            String resolved = o.optString("r", null);
            classMap.put(original, resolved);
        }
    }

    private void parseFields(JSONArray arr) {
        if (arr == null) return;
        for (int i = 0; i < arr.length(); i++) {
            JSONObject o = arr.optJSONObject(i);
            if (o == null) continue;
            String className = o.optString("c", null);
            String original = o.optString("o", null);
            String resolved = o.optString("r", null);
            fieldMap.put(className + "#" + original, resolved);
        }
    }

    private void parseMethods(JSONArray arr) {
        if (arr == null) return;
        for (int i = 0; i < arr.length(); i++) {
            JSONObject o = arr.optJSONObject(i);
            if (o == null) continue;
            String className = o.optString("c", null);
            String original = o.optString("o", null);
            String resolved = o.optString("r", null);
            methodMap.put(className + "#" + original, resolved);
        }
    }

    private void parseMethodAlias(JSONObject obj) {
        if (obj == null) return;
        Iterator<String> keys = obj.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            String value = obj.optString(key, null);
            methodAliasMap.put(key, value);
        }
    }

    public String resolveClass(String original) {
        return classMap.getOrDefault(original, original);
    }

    public String resolveField(String className, String original) {
        return fieldMap.getOrDefault(className + "#" + original, original);
    }

    public String resolveMethod(String className, String original) {
        String aliased = methodAliasMap.getOrDefault(className + "#" + original, original);
        return methodMap.getOrDefault(className + "#" + aliased, aliased);
    }
}

File: app/src/main/java/com/my/televip/obfuscate/struct/ClientObfuscationData.kt
package com.my.televip.obfuscate.struct

import org.json.JSONArray
import org.json.JSONObject

class ClientObfuscationData {

    private val classMap = HashMap<String, String>()
    private val fieldMap = HashMap<String, String>()
    private val methodMap = HashMap<String, String>()
    private val methodAliasMap = HashMap<String, String>()

    private fun parseClasses(arr: JSONArray?) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val original = o.optString("o", null) ?: continue
            val resolved = o.optString("r", null) ?: continue
            classMap[original] = resolved
        }
    }

    private fun parseFields(arr: JSONArray?) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val className = o.optString("c", null) ?: continue
            val original = o.optString("o", null) ?: continue
            val resolved = o.optString("r", null) ?: continue
            fieldMap["$className#$original"] = resolved
        }
    }

    private fun parseMethods(arr: JSONArray?) {
        if (arr == null) return
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val className = o.optString("c", null) ?: continue
            val original = o.optString("o", null) ?: continue
            val resolved = o.optString("r", null) ?: continue
            methodMap["$className#$original"] = resolved
        }
    }

    private fun parseMethodAlias(obj: JSONObject?) {
        if (obj == null) return
        val keys = obj.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = obj.optString(key, null)
            if (value != null) {
                methodAliasMap[key] = value
            }
        }
    }

    fun resolveClass(original: String): String {
        return classMap[original] ?: original
    }

    fun resolveField(className: String, original: String): String {
        return fieldMap["$className#$original"] ?: original
    }

    fun resolveMethod(className: String, original: String): String {
        val aliased = methodAliasMap["$className#$original"] ?: original
        return methodMap["$className#$aliased"] ?: aliased
    }

    companion object {
        @JvmStatic
        fun parse(clientJson: String, clientAliasJson: String): ClientObfuscationData {
            val data = ClientObfuscationData()
            try {
                val root = JSONObject(clientJson)
                data.parseClasses(root.optJSONArray("classes"))
                data.parseFields(root.optJSONArray("fields"))
                data.parseMethods(root.optJSONArray("methods"))
                val aliasRoot = JSONObject(clientAliasJson)
                data.parseMethodAlias(aliasRoot.optJSONObject("methodAlias"))
            } catch (t: Throwable) {
                t.printStackTrace()
            }
            return data
        }
    }
}
